flytekit/extras/tensorflow/records.py

```diff
@@ -0,0 +1,133 @@
+            )
+        )
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE,
+            )
+        )
+        local_dir = ctx.file_access.get_random_local_directory()
+        remote_path = ctx.file_access.get_random_remote_directory()
+        if isinstance(python_val, TFRecordsDirectory):
+            for i, val in enumerate(python_val):
+                local_path = f"{local_dir}/part_{i}"
+                with tf.io.TFRecordWriter(local_path) as writer:
+                    writer.write(val.SerializeToString())
+        elif isinstance(python_val, TFRecordFile):
+            local_path = os.path.join(local_dir, "0000.tfrecord")
+            with tf.io.TFRecordWriter(local_path) as writer:
+                writer.write(python_val.SerializeToString())
+        else:
+            raise AssertionError(
+                f"TensorflowRecordsTransformer can only return TFRecordFile or TFRecordsDirectory types from a task, "
+                f"returned object type provided is {type(python_val)}"
+            )
+        ctx.file_access.upload_directory(local_dir, remote_path)
+        return Literal(scalar=Scalar(blob=Blob(metadata=meta, uri=remote_path)))
+
+    def to_python_value(
+        self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[TFRecordDatasetV2]
+    ) -> TFRecordDatasetV2:
+        try:
+            uri = lv.scalar.blob.uri
+        except AttributeError:
+            TypeTransformerFailedError(f"Cannot convert from {lv} to {expected_python_type}")
+
+        _, metadata = extract_metadata(expected_python_type)
+        local_dir = ctx.file_access.get_random_local_directory()
+        files = os.scandir(uri)
+        filenames = [f.name for f in files if os.path.getsize(f) > 0]
+        ctx.file_access.get_data(uri, local_dir, is_multipart=True if len(filenames) > 1 else False)
+        # load .tfrecord into tf.data.TFRecordDataset
+        return tf.data.TFRecordDataset(
+            filenames=filenames,
+            compression_type=metadata.compression_type,
+            buffer_size=metadata.buffer_size,
+            num_parallel_reads=metadata.num_parallel_read,
+            name=metadata.name,
+        )
+
+    def guess_python_type(self, literal_type: LiteralType) -> Type[T]:
+        if (
+            literal_type.blob is not None
+            and literal_type.blob.dimensionality == _core_types.BlobType.BlobDimensionality.SINGLE
+            and literal_type.blob.format == self.TENSORFLOW_FORMAT
+        ):
+            return T
+
+        raise ValueError(f"Transformer {self} cannot reverse {literal_type}")
+
+
+class TensorflowExampleRecordsTransformer(TensorflowRecordsTransformerBase[example_pb2.Example]):
```

Comment:
I'm guessing the type here needs to be `TFRecordFile` and `TFRecordsDirectory`. You haven't established a link between the flytefile and flytedirectory you defined with this transformer. Also, I'm not sure how not to associate two transformers with `TFRecordFile` — a FlyteFile transformer and this. I'm guessing it just needs to be a `typing.NewType`, not a FlyteFile type. WDYT @cosmicBboy?

Reply:
Do we also want to associate `tf.data.Dataset` with `TFRecordFile` and `TFRecordsDirectory`?

Reply:
I'm a little confused by the implementation of [this spec](https://github.com/flyteorg/flytekit/pull/1240#issuecomment-1292623521)...

I think `TFRecordFile` and `TFRecordsDirectory` should follow the conventions for [flyte files](https://github.com/flyteorg/flytekit/blob/master/flytekit/types/file/__init__.py#L108-L111) and [directories](https://github.com/flyteorg/flytekit/blob/master/flytekit/types/directory/__init__.py#L22), then implement a separate `TFRecordsFileTransformer` and `TFRecordsDirectoryTransformer`, which subclass `FlyteFile` and `FlyteDirectory`, respectively, overriding the `to_python` and `to_literal` methods as necessary to implement the tensorflow-specific implementation.

I wouldn't worry to much about DRY'ing this code by implementing a single transformer for all three of these types... I think a transformer for each of these types will be fine and actually easier to maintain.

Shared functionality can be implemented as helper functions so we don't conflate transformers in a single transformer class implementation.